### PR TITLE
Fixes #22350 - Make REX provider more generic

### PR DESCRIPTION
**Pull request by dLobatog**
This allows to reuse the effective_interfaces/effective_user methods in other provider implementations such as https://github.com/theforeman/foreman_ansible/commit/a5e0827bc3ec6c8ab82f968907857a15646305d5

### CODE REVIEW

File: app/models/remote_execution_provider.rb [hunk 1]
```diff
@@ -27,5 +27,43 @@ def humanized_name
     def supports_effective_user?
       false
     end
+
+    def effective_user(template_invocation)
+      template_invocation.effective_user
+    end
+
+    def effective_user_method(host)
+      method = host_setting(host, :remote_execution_effective_user_method)
```
Comment (adamruzicka) [inline app/models/remote_execution_provider.rb]: The class should probably provide `host_setting` as well

File: app/models/remote_execution_provider.rb [hunk 2]
```diff
@@ -27,5 +27,47 @@ def humanized_name
...
+      unless EFFECTIVE_USER_METHODS.include?(method)
```
Comment (adamruzicka) [inline app/models/remote_execution_provider.rb]: The class should provide the `EFFECTIVE_USER_METHODS` constant as well.

```
2018-01-23T10:28:58 6fe85da4 [background] [E] uninitialized constant EFFECTIVE_USER_METHODS (NameError)
 | /home/aruzicka/git/foreman-remote-execution/app/models/remote_execution_provider.rb:37:in `effective_user_method'
 | /home/aruzicka/git/foreman-remote-execution/app/models/ssh_execution_provider.rb:9:in `proxy_command_options'                                   
 | /home/aruzicka/git/foreman-remote-execution/app/lib/actions/remote_execution/run_host_job.rb:51:in `plan'
```

**Comment by adamruzicka** [comment]
Tests running here http://ci.theforeman.org/job/test_plugin_pull_request/4549/

**Comment by adamruzicka** [comment]
Test failure seems unrelated.

**Comment by iNecas** [comment]
Thanks @dLobatog and @adamruzicka

**Pull request merged**